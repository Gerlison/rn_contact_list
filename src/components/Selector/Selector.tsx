import React, { memo, useMemo, useState } from 'react';
import { PickerProps } from 'react-native';
import styled from 'styled-components/native';
import { Picker } from '@react-native-picker/picker';

import Text from '../Text';

import icArrow from '../../../assets/icons/icArrow.png';

interface Props extends PickerProps {
  isLoading: boolean;
  items: {
    label: string;
    value: string;
  }[];
}

const Selector = ({ items, isLoading, ...props }: Props): JSX.Element => {
  const [isOpen, setIsOpen] = useState(false);

  const renderPickerItems = useMemo(() => {
    if (isLoading)
      return (
        <Picker.Item
          testID="selector loading"
          color="#a3a3a3"
          value="loading"
          label="Loading..."
        />
      );

    return items.map((item) => (
      <Picker.Item
        testID="selector item"
        key={item.value}
        label={item.label}
        value={item.value}
      />
    ));
  }, [isLoading, items]);

  return (
    <>
      {isOpen && (
        <S.Backdrop
          testID="selector backdrop"
          onPress={() => setIsOpen(false)}
        />
      )}
      <S.Container testID="selector">
        <S.Touchable
          testID="selector touchable"
          isOpen={isOpen}
          onPress={() => setIsOpen((old) => !old)}
        >
          <Text
            size={14}
            variant="Poppins-Medium"
            color={props.selectedValue ? '#000' : 'rgba(0,0,0,.5)'}
          >
            {props.selectedValue || 'City'}
          </Text>
          <S.Image isOpen={isOpen} source={icArrow} />
        </S.Touchable>
        {isOpen && (
          <Picker
            testID="selector picker"
            enabled={!isLoading}
            itemStyle={ITEM_STYLE}
            {...props}
          >
            {renderPickerItems}
          </Picker>
        )}
      </S.Container>
    </>
  );
};

const S = {
  Backdrop: styled.TouchableOpacity`
    position: absolute;
    top: 0px;
    bottom: 0px;
    left: 0px;
    right: 0px;
  `,
  Container: styled.View`
    width: 100%;
    margin-bottom: 24px;
    border-radius: 20px;
    border-color: #aeaeae;
    border-width: 1px;
  `,
  Touchable: styled.TouchableOpacity<{ isOpen: boolean }>`
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    padding: 20px;
    padding-bottom: ${({ isOpen }) => (isOpen ? 0 : 20)}px;
  `,
  Image: styled.Image.attrs({
    resizeMode: 'contain',
  })<{ isOpen: boolean }>`
    height: 12px;
    transform: ${({ isOpen }) => `rotate(${isOpen ? 180 : 0}deg)`};
  `,
};

const ITEM_STYLE = { fontFamily: 'Poppins-Medium', fontSize: 16 };

export default memo(Selector);
