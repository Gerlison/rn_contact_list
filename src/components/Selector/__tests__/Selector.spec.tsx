import React from 'react';
import { act, fireEvent, render } from '@testing-library/react-native';

import Selector from '../Selector';

const mockedItems = [
  {
    label: 'test',
    value: 'test',
  },
  {
    label: 'test2',
    value: 'test2',
  },
];

describe('Selector', () => {
  it('SHOULD render correctly', () => {
    const component = render(
      <Selector
        items={[]}
        isLoading={false}
        onValueChange={jest.fn()}
        selectedValue={'test'}
      />,
    );
    expect(component).toBeDefined();
  });

  it('SHOULD match snapshot', () => {
    const component = render(
      <Selector
        items={[]}
        isLoading={false}
        onValueChange={jest.fn()}
        selectedValue={'test'}
      />,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('SHOULD show picker on click touchable', async () => {
    const component = render(
      <Selector
        items={mockedItems}
        isLoading={false}
        onValueChange={jest.fn()}
        selectedValue=""
      />,
    );

    act(() => {
      fireEvent.press(component.getByTestId('selector touchable'));
    });

    expect(component.queryByTestId('selector picker')).toBeTruthy();
  });
});
