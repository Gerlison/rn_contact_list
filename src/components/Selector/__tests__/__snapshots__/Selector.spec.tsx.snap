// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector SHOULD match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "borderColor": "#aeaeae",
        "borderRadius": 20,
        "borderWidth": 1,
        "marginBottom": 24,
        "width": "100%",
      },
    ]
  }
  testID="selector"
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "opacity": 1,
        "paddingBottom": 20,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 20,
      }
    }
    testID="selector touchable"
  >
    <Text
      color="#000"
      family="Poppins-Medium"
      size={14}
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "Poppins-Medium",
            "fontSize": 14,
          },
        ]
      }
    >
      test
    </Text>
    <Image
      isOpen={false}
      resizeMode="contain"
      source={1}
      style={
        Array [
          Object {
            "height": 12,
            "transform": Array [
              Object {
                "rotate": "0deg",
              },
            ],
          },
        ]
      }
    />
  </View>
</View>
`;
